### PR TITLE
Deprecate --volume-driver

### LINKED ISSUE #16069: Deprecate "--volume-driver" option
The `--volume-driver` option was never a good fit; this option can only be specified _once_, which isn't really useful if a container uses multiple volumes (and different drivers need to be used for each one)

Now that `volume create` command has been merged, I think we should deprecate `--volume-driver` in favor of `volume create`, i.e.:

**old approach**

``` bash
docker run -d -v volumename:/data --volume-driver=flocker myimage
```

**new approach**

``` bash
docker run -d \
    -v $(docker volume create --driver=flocker --name=volumename):/data \
    -v $(docker volume create --driver=foobar --name=something):/somewhere \
    myimage
```

I'd like to have this decided on for the 1.9 release, to give a clear migration path for users that currently rely on this feature.

### PR DESCRIPTION
**Pull request by erikh**
/cc @cpuguy83 @calavera @srust 

This deprecates the `--volume-driver` flag and pre-emptively schedules it for removal in 1.11. Please let me know if any wording needs to be adjusted, as this is principally a docs patch.

Thanks!

-Erik

Fixes #16069

### CODE REVIEW

**Comment by jfrazelle** [comment]
gofmt

**Comment by erikh** [comment]
my save hook wasn’t working, thanks, fixed.

**Comment by cpuguy83** [comment]
It would be good to do this on the daemon side as well (in the daemon logs).
Also there is an open issue for this by @thajeztah, but I'm on my phone right now so can't find it.

**Comment by erikh** [comment]
I’ll dig it up, don’t worry about it.

**Comment by erikh** [comment]
This should be available now, PTAL

File: docs/misc/deprecated.md [hunk 1]
```diff
@@ -22,6 +22,9 @@ variants:
 
     docker run -c (--cpu-shares)
 
+`docker run --volume-driver` is now deprecated in favor of `docker volume
```
Comment (thaJeztah) [inline docs/misc/deprecated.md]: Early docs review; This needs its own head, right now it's below "command line short variant options"

**Comment by thaJeztah** [comment]
Thanks, Erik

**Comment by jhowardmsft** [comment]
Could we make this a daemon-side check rather than client? That way, on the Windows path, we could just error out if supplied. Just thinking ahead to remove any possible baggage on the Windows volume support (#16433) which will likely be deprecated by the time we RTM Server 2016.

**Comment by erikh** [comment]
It should be both as of the latest revision.

**Comment by erikh** [comment]
Oh, error. I was told to make it warn for now.
-Erik

**Comment by cpuguy83** [comment]
major +1 from me

**Comment by erikh** [comment]
Can you clarify what I should be doing here, if anything?

**Comment by calavera** [comment]
such a short life :crying_cat_face: 

I'm conflicted because I think it's a really nice "one-time-only" thing. But I'll be okay if other maintainers want to deprecate it.

**Comment by vdemeester** [comment]
> I'm conflicted because I think it's a really nice "one-time-only" thing. But I'll be okay if other maintainers want to deprecate it.

What do you mean by "one-time-only" ? It also feels strange to have `--volume-driver` and a `volume` command at the same time though.

**Comment by calavera** [comment]
I know it feels weird to have both. I'm conflicted because I like to do this for certain cases:

```
$ docker run --rm --volume-driver --volume foo foo:/bar ...
```

Rather than:

```
$ docker volume create --driver foo --name foo
$ docker run --rm --volume foo:/bar ...
```

**Comment by calavera** [comment]
but I can learn to leave without it.

**Comment by thaJeztah** [comment]
Yes, I can understand why you like the simplicity of --volume-driver, I just think it quickly becomes complicated of multiple volumes are involved (including volumes defined in the image).

So apologies for "killing" the flag, but I hope you can understand my reasoning

**Comment by jfrazelle** [comment]
i kinida like the flag, but im not married to it so either way

**Comment by vdemeester** [comment]
all right then, I'll drop the ball :-) : LGTM :wink:

**Comment by calavera** [comment]
> I just think it quickly becomes complicated of multiple volumes are involved

that doesn't imply that we have to kill this flag. We have a whole new api to manage those complicated cases.

I see this flag more as a shortcut for simple cases than an impediment for complicated cases.

**Comment by jfrazelle** [comment]
ok so as to get this show on the road it seems like we are all split and until we have concensus lets keep the flag.

**Comment by lukemarsden** [comment]
I am -1 on this because it breaks our integration with tools that use `docker run` as an integration point.

**Comment by thaJeztah** [comment]
@lukemarsden -1 on removing the flag, or -1 on closing this PR?

**Comment by lukemarsden** [comment]
-1 on removing the flag. Thanks for clarifying!

**Comment by lukemarsden** [comment]
Also, until we have a docker compose story with first class volumes, we should keep `volume_driver` here and there, for consistency, IMHO :)

**Comment by thaJeztah** [comment]
Cool, thanks @lukemarsden

**Comment by osallou** [comment]
If you create a volume first, how do you handle multi node case?

    $ docker volume create --driver foo --name foo
    $ docker run --rm --volume foo:/bar ...

With swarm, mesos etc.., container will be scheduled based on requirements (ram, cpu, ...) on a node. If we create the volume first, which docker instance will handle the storage plugin?

docker volume create will call the VolumeDriver.Create on one node then
docker run.. --volume foo will call the VolumeDriver.Mount/Unmount on the node selected by swarm/... ?

**Comment by cpuguy83** [comment]
@osallou swarm does the create on every node.

**Pull request closed without merging**